File: src/lib/services/aggregator.ts
import { getAllSources } from "@/lib/sources";
import type {
  CurrentData,
  HourlyData,
  DailyData,
  WeatherSource,
} from "@/lib/sources";
import { getConditionIcon } from "@/lib/sources/base";
import type {
  SourceData,
  AgreementInfo,
  CurrentWeather,
  HourlyForecast,
  DailyForecast,
  TodaySourceTemp,
} from "@/lib/types";

interface FetchResult {
  name: string;
  data: CurrentData | HourlyData[] | DailyData[] | null;
  elapsed: number;
  ok: boolean;
}

async function fetchWithTiming(
  source: WeatherSource,
  method: "fetchCurrent" | "fetchHourly" | "fetchDaily",
  lat: number,
  lon: number,
): Promise<FetchResult> {
  const start = Date.now();
  try {
    const timeoutPromise = new Promise<never>((_, reject) =>
      setTimeout(() => reject(new Error("Timeout")), 4000),
    );
    const data = await Promise.race([source[method](lat, lon), timeoutPromise]);
    return {
      name: source.name,
      data: data as CurrentData,
      elapsed: Date.now() - start,
      ok: true,
    };
  } catch {
    return {
      name: source.name,
      data: null,
      elapsed: Date.now() - start,
      ok: false,
    };
  }
}

function calculateConfidence(
  sourceTemps: Record<string, number>,
): Record<string, number> {
  const values = Object.values(sourceTemps);
  if (values.length < 2) {
    return Object.fromEntries(Object.keys(sourceTemps).map((k) => [k, 95]));
  }

  const sorted = [...values].sort((a, b) => a - b);
  const median = sorted[Math.floor(sorted.length / 2)];
  const maxSpread = Math.max(...values) - Math.min(...values);

  const scores: Record<string, number> = {};
  for (const [name, temp] of Object.entries(sourceTemps)) {
    const deviation = Math.abs(temp - median);
    let score = Math.max(40, 100 - deviation * 15);
    if (maxSpread < 2) score = Math.min(100, score + 5);
    scores[name] = Math.round(score * 10) / 10;
  }
  return scores;
}

function calculateAgreement(
  sourceTemps: Record<string, number>,
): AgreementInfo {
  const values = Object.values(sourceTemps);
  if (values.length < 2) {
    return {
      level: "necunoscut",
      color: "gray",
      max_deviation: 0,
      description: "Insuficiente surse pentru comparație",
    };
  }

  const maxDev =
    Math.round((Math.max(...values) - Math.min(...values)) * 10) / 10;

  if (maxDev < 2)
    return {
      level: "puternic",
      color: "green",
      max_deviation: maxDev,
      description: `Sursele sunt de acord (diferență: ${maxDev}°C)`,
    };
  if (maxDev < 5)
    return {
      level: "moderat",
      color: "yellow",
      max_deviation: maxDev,
      description: `Acord moderat între surse (diferență: ${maxDev}°C)`,
    };
  return {
    level: "dezacord",
    color: "red",
    max_deviation: maxDev,
    description: `Dezacord între surse (diferență: ${maxDev}°C)`,
  };
}

function weightedAverage(
  temps: Record<string, number>,
  confidence: Record<string, number>,
): number {
  let totalWeight = 0;
  let weightedSum = 0;
  for (const [name, temp] of Object.entries(temps)) {
    const w = confidence[name] || 50;
    weightedSum += temp * w;
    totalWeight += w;
  }
  return totalWeight === 0
    ? 0
    : Math.round((weightedSum / totalWeight) * 10) / 10;
}

function majorityCondition(conditions: Record<string, string>): string {
  const counts: Record<string, number> = {};
  for (const c of Object.values(conditions)) {
    counts[c] = (counts[c] || 0) + 1;
  }
  let best = "necunoscut";
  let bestCount = 0;
  for (const [c, count] of Object.entries(counts)) {
    if (count > bestCount) {
      best = c;
      bestCount = count;
    }
  }
  return best;
}

export async function aggregateCurrent(
  lat: number,
  lon: number,
): Promise<{
  current: CurrentWeather;
  comparison: SourceData[];
  agreement: AgreementInfo;
  avgConfidence: number;
}> {
  const sources = getAllSources();
  const results = await Promise.all(
    sources.map((s) => fetchWithTiming(s, "fetchCurrent", lat, lon)),
  );

  const sourceTemps: Record<string, number> = {};
  const sourceConditions: Record<string, string> = {};
  const sourceHumidity: Record<string, number> = {};
  const sourceWind: Record<string, number> = {};

  for (const r of results) {
    if (r.ok && r.data) {
      const d = r.data as CurrentData;
      sourceTemps[r.name] = d.temperature;
      sourceConditions[r.name] = d.condition;
      sourceHumidity[r.name] = d.humidity;
      sourceWind[r.name] = d.wind_speed;
    }
  }

  const confidence = calculateConfidence(sourceTemps);
  const agreement = calculateAgreement(sourceTemps);

  const comparison: SourceData[] = results.map((r) => {
    const d = r.ok ? (r.data as CurrentData) : null;
    return {
      source: r.name,
      temperature: d?.temperature ?? null,
      humidity: d?.humidity ?? null,
      wind_speed: d?.wind_speed ?? null,
      condition: d?.condition ?? null,
      available: r.ok && d !== null,
      confidence: confidence[r.name] || 0,
      response_time_ms: r.elapsed,
    };
  });

  const condition = majorityCondition(sourceConditions);
  let windDir = 0;
  let precip = 0;
  for (const r of results) {
    if (r.ok && r.data) {
      windDir = (r.data as CurrentData).wind_direction;
      precip = (r.data as CurrentData).precipitation;
      break;
    }
  }

  const current: CurrentWeather = {
    temperature: weightedAverage(sourceTemps, confidence),
    humidity: Math.round(
      weightedAverage(
        Object.fromEntries(
          Object.entries(sourceHumidity).map(([k, v]) => [k, v]),
        ),
        confidence,
      ),
    ),
    wind_speed: Math.round(weightedAverage(sourceWind, confidence) * 10) / 10,
    wind_direction: windDir,
    precipitation: precip,
    condition,
    condition_icon: getConditionIcon(condition),
  };

  const confValues = Object.values(confidence);
  const avgConfidence = confValues.length
    ? Math.round(
        (confValues.reduce((a, b) => a + b, 0) / confValues.length) * 10,
      ) / 10
    : 0;

  return { current, comparison, agreement, avgConfidence };
}

export async function aggregateHourly(
  lat: number,
  lon: number,
): Promise<HourlyForecast[]> {
  const sources = getAllSources();
  const results = await Promise.all(
    sources.map((s) => fetchWithTiming(s, "fetchHourly", lat, lon)),
  );

  // Open-Meteo ca baza
  let primary: HourlyData[] | null = null;
  for (const r of results) {
    if (r.name === "open_meteo" && r.ok && r.data) {
      primary = r.data as HourlyData[];
      break;
    }
  }
  if (!primary) {
    for (const r of results) {
      if (r.ok && r.data) {
        primary = r.data as HourlyData[];
        break;
      }
    }
  }

  return (primary || []).map((h) => ({
    hour: h.hour,
    timestamp: h.timestamp,
    temperature: h.temperature,
    humidity: h.humidity,
    wind_speed: h.wind_speed,
    precipitation: h.precipitation,
    condition: h.condition,
  }));
}

const SOURCE_LABELS: Record<string, string> = {
  open_meteo: "Open-Meteo",
  ecmwf: "ECMWF",
  openweather: "OpenWeather",
  weatherapi: "WeatherAPI",
};

export async function aggregateDaily(
  lat: number,
  lon: number,
): Promise<{ days: DailyForecast[]; todaySources: TodaySourceTemp[] }> {
  const sources = getAllSources();
  const results = await Promise.all(
    sources.map((s) => fetchWithTiming(s, "fetchDaily", lat, lon)),
  );

  // Pastreaza sursa pentru fiecare DailyData
  const daysData: Record<string, { source: string; data: DailyData }[]> = {};
  for (const r of results) {
    if (r.ok && r.data) {
      for (const day of r.data as DailyData[]) {
        if (!daysData[day.date]) daysData[day.date] = [];
        daysData[day.date].push({ source: r.name, data: day });
      }
    }
  }

  const sortedDates = Object.keys(daysData).sort().slice(0, 7);

  // Calculeaza per-sursa pentru ziua de azi (prima zi)
  let todaySources: TodaySourceTemp[] = [];
  if (sortedDates.length > 0) {
    const todayItems = daysData[sortedDates[0]];
    const todayMaxTemps: Record<string, number> = {};
    for (const item of todayItems) {
      todayMaxTemps[item.source] = item.data.temp_max;
    }
    const todayConfidence = calculateConfidence(todayMaxTemps);

    todaySources = todayItems.map((item) => ({
      source: item.source,
      label: SOURCE_LABELS[item.source] || item.source,
      temp_min: item.data.temp_min,
      temp_max: item.data.temp_max,
      confidence: todayConfidence[item.source] || 50,
    }));
  }

  const days = sortedDates.map((dateStr) => {
    const items = daysData[dateStr];

    // Calculeaza confidence separat pentru temp_max si temp_min
    // (o sursa poate fi outlier doar pe una dintre ele)
    const maxTemps: Record<string, number> = {};
    const minTemps: Record<string, number> = {};
    for (const item of items) {
      maxTemps[item.source] = item.data.temp_max;
      minTemps[item.source] = item.data.temp_min;
    }
    const confidenceMax = calculateConfidence(maxTemps);
    const confidenceMin = calculateConfidence(minTemps);

    const conditions = items.map((i) => i.data.condition);
    const condCounts: Record<string, number> = {};
    for (const c of conditions) condCounts[c] = (condCounts[c] || 0) + 1;
    let bestCond = "necunoscut";
    let bestCount = 0;
    for (const [c, n] of Object.entries(condCounts)) {
      if (n > bestCount) {
        bestCond = c;
        bestCount = n;
      }
    }

    // Media humidity din sursele care o ofera
    const humidities = items
      .map((i) => i.data.humidity)
      .filter((h): h is number => typeof h === "number" && h > 0);
    const avgHumidity =
      humidities.length > 0
        ? Math.round(humidities.reduce((a, b) => a + b, 0) / humidities.length)
        : 0;

    return {
      date: dateStr,
      day_name: items[0].data.day_name,
      temp_min: weightedAverage(minTemps, confidenceMin),
      temp_max: weightedAverage(maxTemps, confidenceMax),
      humidity: avgHumidity,
      precipitation:
        Math.round(Math.max(...items.map((i) => i.data.precipitation)) * 10) /
        10,
      wind_speed:
        Math.round(Math.max(...items.map((i) => i.data.wind_speed)) * 10) / 10,
      condition: bestCond,
    };
  });

  return { days, todaySources };
}

File: backend/services/aggregator.py
import asyncio
import time
from statistics import median
from sources import OpenMeteoSource, OpenWeatherSource, WeatherAPISource, ECMWFSource
from models import (
    SourceData, AgreementInfo, CurrentWeather, HourlyForecast, DailyForecast,
)
from sources.base import get_condition_icon


ALL_SOURCES = [
    OpenMeteoSource(),
    OpenWeatherSource(),
    WeatherAPISource(),
    ECMWFSource(),
]


async def _fetch_with_timing(source, method: str, lat: float, lon: float) -> tuple[str, dict | list | None, int, bool]:
    """Fetch de la o sursa cu masurare timp si tratare erori."""
    start = time.monotonic()
    try:
        func = getattr(source, method)
        data = await func(lat, lon)
        elapsed = int((time.monotonic() - start) * 1000)
        return source.name, data, elapsed, True
    except Exception:
        elapsed = int((time.monotonic() - start) * 1000)
        return source.name, None, elapsed, False


def _calculate_confidence(source_temps: dict[str, float]) -> dict[str, float]:
    """Calculeaza scorul de incredere per sursa bazat pe deviatie fata de mediana."""
    if not source_temps:
        return {}

    values = list(source_temps.values())
    if len(values) < 2:
        return {name: 95.0 for name in source_temps}

    med = median(values)
    max_spread = max(values) - min(values) if len(values) > 1 else 0

    scores = {}
    for name, temp in source_temps.items():
        deviation = abs(temp - med)
        # Scor de baza 100, scade cu deviatie
        score = max(40.0, 100.0 - (deviation * 15.0))
        # Bonus daca spread-ul e mic (sursele sunt de acord)
        if max_spread < 2:
            score = min(100.0, score + 5.0)
        scores[name] = round(score, 1)

    return scores


def _calculate_agreement(source_temps: dict[str, float]) -> AgreementInfo:
    """Calculeaza nivelul de acord intre surse."""
    if len(source_temps) < 2:
        return AgreementInfo(
            level="necunoscut",
            color="gray",
            max_deviation=0,
            description="Insuficiente surse pentru comparație",
        )

    values = list(source_temps.values())
    max_dev = round(max(values) - min(values), 1)

    if max_dev < 2:
        return AgreementInfo(
            level="puternic",
            color="green",
            max_deviation=max_dev,
            description=f"Sursele sunt de acord (diferență: {max_dev}°C)",
        )
    elif max_dev < 5:
        return AgreementInfo(
            level="moderat",
            color="yellow",
            max_deviation=max_dev,
            description=f"Acord moderat între surse (diferență: {max_dev}°C)",
        )
    else:
        return AgreementInfo(
            level="dezacord",
            color="red",
            max_deviation=max_dev,
            description=f"Dezacord între surse (diferență: {max_dev}°C)",
        )


def _weighted_average(source_temps: dict[str, float], confidence: dict[str, float]) -> float:
    """Media ponderata cu scoruri de incredere."""
    total_weight = 0
    weighted_sum = 0
    for name, temp in source_temps.items():
        weight = confidence.get(name, 50.0)
        weighted_sum += temp * weight
        total_weight += weight
    if total_weight == 0:
        return 0
    return round(weighted_sum / total_weight, 1)


def _majority_condition(source_conditions: dict[str, str]) -> str:
    """Conditia meteo prin vot majoritar."""
    if not source_conditions:
        return "necunoscut"
    counts: dict[str, int] = {}
    for cond in source_conditions.values():
        counts[cond] = counts.get(cond, 0) + 1
    return max(counts, key=counts.get)


async def aggregate_current(lat: float, lon: float) -> tuple[CurrentWeather, list[SourceData], AgreementInfo, float]:
    """Agregheaza datele curente din toate sursele."""
    tasks = [_fetch_with_timing(s, "fetch_current", lat, lon) for s in ALL_SOURCES]
    results = await asyncio.gather(*tasks)

    source_temps = {}
    source_conditions = {}
    source_humidity = {}
    source_wind = {}
    comparison = []

    for name, data, elapsed, ok in results:
        if ok and data:
            source_temps[name] = data["temperature"]
            source_conditions[name] = data["condition"]
            source_humidity[name] = data["humidity"]
            source_wind[name] = data["wind_speed"]

    confidence = _calculate_confidence(source_temps)
    agreement = _calculate_agreement(source_temps)

    for name, data, elapsed, ok in results:
        comparison.append(SourceData(
            source=name,
            temperature=data["temperature"] if ok and data else None,
            humidity=data["humidity"] if ok and data else None,
            wind_speed=data["wind_speed"] if ok and data else None,
            condition=data["condition"] if ok and data else None,
            available=ok and data is not None,
            confidence=confidence.get(name, 0),
            response_time_ms=elapsed,
        ))

    # Prognoza agregata ponderata
    avg_temp = _weighted_average(source_temps, confidence) if source_temps else 0
    avg_humidity = int(_weighted_average(
        {k: float(v) for k, v in source_humidity.items()}, confidence
    )) if source_humidity else 0
    avg_wind = round(_weighted_average(
        source_wind, confidence
    ), 1) if source_wind else 0
    condition = _majority_condition(source_conditions)

    # Wind direction din prima sursa disponibila
    wind_dir = 0
    precip = 0.0
    for _, data, _, ok in results:
        if ok and data:
            wind_dir = data.get("wind_direction", 0)
            precip = data.get("precipitation", 0)
            break

    current = CurrentWeather(
        temperature=avg_temp,
        humidity=avg_humidity,
        wind_speed=avg_wind,
        wind_direction=wind_dir,
        precipitation=precip,
        condition=condition,
        condition_icon=get_condition_icon(condition),
    )

    avg_confidence = round(sum(confidence.values()) / len(confidence), 1) if confidence else 0
    return current, comparison, agreement, avg_confidence


async def aggregate_hourly(lat: float, lon: float) -> list[HourlyForecast]:
    """Agregheaza prognoza orara (sursa principala: Open-Meteo)."""
    tasks = [_fetch_with_timing(s, "fetch_hourly", lat, lon) for s in ALL_SOURCES]
    results = await asyncio.gather(*tasks)

    # Folosim Open-Meteo ca baza (are cele mai multe ore)
    primary = None
    for name, data, _, ok in results:
        if name == "open_meteo" and ok and data:
            primary = data
            break

    if not primary:
        for _, data, _, ok in results:
            if ok and data:
                primary = data
                break

    if not primary:
        return []

    forecasts = []
    for item in primary:
        forecasts.append(HourlyForecast(
            hour=item["hour"],
            timestamp=item["timestamp"],
            temperature=item["temperature"],
            humidity=item["humidity"],
            wind_speed=item["wind_speed"],
            precipitation=item.get("precipitation", 0),
            condition=item["condition"],
        ))
    return forecasts


async def aggregate_daily(lat: float, lon: float) -> list[DailyForecast]:
    """Agregheaza prognoza zilnica din toate sursele."""
    tasks = [_fetch_with_timing(s, "fetch_daily", lat, lon) for s in ALL_SOURCES]
    results = await asyncio.gather(*tasks)

    # Combina zilele din toate sursele
    days_data: dict[str, list[dict]] = {}
    for name, data, _, ok in results:
        if ok and data:
            for day in data:
                date_key = day["date"]
                if date_key not in days_data:
                    days_data[date_key] = []
                days_data[date_key].append(day)

    forecasts = []
    for date_str in sorted(days_data.keys())[:7]:
        items = days_data[date_str]
        temp_mins = [d["temp_min"] for d in items if d.get("temp_min") is not None]
        temp_maxs = [d["temp_max"] for d in items if d.get("temp_max") is not None]
        conditions = [d["condition"] for d in items if d.get("condition")]

        condition_counts: dict[str, int] = {}
        for c in conditions:
            condition_counts[c] = condition_counts.get(c, 0) + 1
        best_condition = max(condition_counts, key=condition_counts.get) if condition_counts else "necunoscut"

        forecasts.append(DailyForecast(
            date=date_str,
            day_name=items[0].get("day_name", ""),
            temp_min=round(sum(temp_mins) / len(temp_mins), 1) if temp_mins else 0,
            temp_max=round(sum(temp_maxs) / len(temp_maxs), 1) if temp_maxs else 0,
            precipitation=round(max((d.get("precipitation", 0) for d in items), default=0), 1),
            wind_speed=round(max((d.get("wind_speed", 0) for d in items), default=0), 1),
            condition=best_condition,
        ))
    return forecasts
